pkg/common/scrubber.go

```diff
@@ -0,0 +1,341 @@
+/*
+Copyright 2023 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/gosecretive"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	ReferencePrefix                  = "$ref:"
+	ReferenceToEnvVarPrefix          = "NUCLIO_B64_"
+	NuclioFlexVolumeSecretNamePrefix = "nuclio-flexvolume"
+	SecretTypeFunctionConfig         = "nuclio.io/functionconfig"
+	SecretTypeV3ioFuse               = "v3io/fuse"
+	SecretContentKey                 = "content"
+	FunctionSecretMountPath          = "/etc/nuclio/secrets"
+)
+
+type Scrubber interface {
+	ValidateReference(objectToScrub interface{},
+		existingSecretMap map[string]string,
+		fieldPath,
+		secretKey,
+		stringValue string) error
+	ConvertMapToConfig(mapConfig interface{}) (interface{}, error)
+}
+
+type AbstractScrubber struct {
+	SensitiveFields []*regexp.Regexp
+	KubeClientSet   kubernetes.Interface
+	ReferencePrefix string
+	Scrubber        Scrubber
+}
+
+// NewAbstractScrubber returns a new AbstractScrubber
+// If the scrubber is only used for restoring, the arguments can be nil
+func NewAbstractScrubber(sensitiveFields []*regexp.Regexp, kubeClientSet kubernetes.Interface) *AbstractScrubber {
+	return &AbstractScrubber{
+		SensitiveFields: sensitiveFields,
+		KubeClientSet:   kubeClientSet,
+		ReferencePrefix: ReferencePrefix,
+	}
+}
+
+// Scrub scrubs sensitive data from a function config
+func (s *AbstractScrubber) Scrub(objectToScrub interface{},
```

Comment:
If you want to make this general to both functionConfig and APIGateways, I would modify all comments and variable names to the more general sense.
e.g `config` instead of `functionConfig` etc.
This relates to all of the functions in this file.